portia/introspection_agents/default_introspection_agent.py

```diff
@@ -0,0 +1,91 @@
+"""The default introspection agent.
+
+This agent looks at the state of a plan run between steps
+and makes decisions about whether execution should continue.
+"""
+
+from langchain.prompts import ChatPromptTemplate, HumanMessagePromptTemplate
+from langchain.schema import SystemMessage
+
+from portia.config import INTROSPECTION_MODEL_KEY, Config
+from portia.introspection_agents.introspection_agent import (
+    BaseIntrospectionAgent,
+    PreStepIntrospection,
+)
+from portia.llm_wrapper import LLMWrapper
+from portia.plan import Plan
+from portia.plan_run import PlanRun
+
+
+class DefaultIntrospectionAgent(BaseIntrospectionAgent):
+    """Default Introspection Agent.
+
+    Implements the BaseIntrospectionAgent interface using an LLM to make decisions about what to do.
+
+    Attributes:
+        config (Config): Configuration settings for the DefaultIntrospectionAgent.
+
+    """
+
+    def __init__(self, config: Config) -> None:
+        """Initialize the DefaultIntrospectionAgent with configuration.
+
+        Args:
+            config (Config): The configuration to initialize the DefaultIntrospectionAgent.
+
+        """
+        self.config = config
+
+        self.prompt = ChatPromptTemplate.from_messages(
+            [
+                SystemMessage(
+                    content=(
+                        "You are a highly skilled reviewer who reviews in flight plan execution."
+                        "Your job is to examine the state of a plan execution (PlanRun) and "
+                        "decide what action should be taken next."
+                        "You should use the current_step_index field to identify the current step"
+                        "in the plan, and the PlanRun state to know what has happened so far."
+                        "The actions that can be taken next in priority order are:"
+                        " - STOP -> stop execution and return the result so far."
+                        " - SKIP -> skip the next step execution."
+                        " - FAIL -> stop execution entirely."
+                        " - CONTINUE -> Continue execution of the next step."
+                        "You should choose an outcome based on the following logic in order:\n"
+                        " - If all remaining steps can not be executed because of the"
```

Comment:
I'm not sure this is quite correct? We should STOP because its done right?

```suggestion
                        " - If the overarching goal of the plan has already been met return SKIP"
```

Reply:
Yea, happy to give it a try. I'm a bit skiptical on doing it overall all goal, as I'm afraid of hallucinating and just stops prematurely.